Data Understanding and Cleaning

In [4]:
import pandas as pd
import numpy as np 

In [5]:
df=pd.read_csv("C:/Users/aryap/Desktop/Projects/Customer_segmentation/data/online_retail.csv")

In [8]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [9]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [7]:
df.shape

(541909, 8)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [12]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

Each row in the dataset represents a single product purchased as part of an invoice. A single invoice may contain multiple rows if multiple products are bought together.

In [ ]:
df.isnull().sum() #see which columns have null values and the sum of all the null values in that column

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

CustomerID has null values so it indicates that there are transactions where customer data is not recorded

In [16]:
#check if there are duplicate rows/entries
df.duplicated().sum()

np.int64(5268)

In [17]:
df[df.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

                InvoiceDate  UnitPrice  CustomerID         Country  
517     2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
527     2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
537     2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
539     2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
555     2010-12-01 11:49:00       2.95     17920.0  United Kingdom  
...                     ...        ...         ...             ...  
541675  2011-12-09 11:34:00       0.39     14446.0  United Kingdom  
541689  2011-12-09 11:34:00       2.49     14446.0  United Kingdom  
541692  2011-12-09 11:34:00       1.95     14446.0  United Kingdom  
541699  2011-12-09 11:34:00       2.10     14446.0  United Kingdom  
541701  2011-12-09 11:34:00       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

In [21]:
# check unique customers
df['CustomerID'].nunique()

4372

In [22]:
df['InvoiceNo'].nunique()

25900

This dataset contains 4372 unique customers and 25900 unique invoices


In [24]:
# sanity check for quantity and price
df['Quantity'].describe()
df['UnitPrice'].describe()

count    541909.000000
mean          4.611114
std          96.759853
min      -11062.060000
25%           1.250000
50%           2.080000
75%           4.130000
max       38970.000000
Name: UnitPrice, dtype: float64

# Handling Missing CustomerID

In [8]:
# number of rows having missing CustomerID
df['CustomerID'].isnull().sum()

np.int64(135080)

In [9]:
# Affecting percentage
df['CustomerID'].isnull().mean()*100

np.float64(24.926694334288598)

In [10]:
#remove rows with missing CustomerID
df=df.dropna(subset=['CustomerID'])

In [11]:
#confirming the removal
df['CustomerID'].isnull().sum()

np.int64(0)

In [16]:
#Convert customerid to integer as it represents an id not a float
df['CustomerID']=df['CustomerID'].astype(int)


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 406829 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    406829 non-null  object 
 1   StockCode    406829 non-null  object 
 2   Description  406829 non-null  object 
 3   Quantity     406829 non-null  int64  
 4   InvoiceDate  406829 non-null  object 
 5   UnitPrice    406829 non-null  float64
 6   CustomerID   406829 non-null  int64  
 7   Country      406829 non-null  object 
 8   CustomerId   406829 non-null  int64  
dtypes: float64(1), int64(3), object(5)
memory usage: 31.0+ MB


In [19]:
del df['CustomerId']

In [20]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 406829 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    406829 non-null  object 
 1   StockCode    406829 non-null  object 
 2   Description  406829 non-null  object 
 3   Quantity     406829 non-null  int64  
 4   InvoiceDate  406829 non-null  object 
 5   UnitPrice    406829 non-null  float64
 6   CustomerID   406829 non-null  int64  
 7   Country      406829 non-null  object 
dtypes: float64(1), int64(2), object(5)
memory usage: 27.9+ MB


**Removing Cancelled Transactions**
- cancelled invoices represent returned or refunded purchases
- if we keep these cancelled transactions, revenue calculations and RFM will be wrong
- only keep completed purchases


In [23]:
# check how many cancelled invoices exist (Cancelled invoices start with C)
df['InvoiceNo'].astype(str).str.startswith('C').sum()

np.int64(8905)

In [24]:
# remove the cancelled invoices
df=df[~df['InvoiceNo'].astype(str).str.startswith('C')]

In [25]:
#confirm the removal
df['InvoiceNo'].astype(str).str.startswith('C').sum()

np.int64(0)

Removing invalid quantity and price values


In [30]:
# check for invalid quantities
(df['Quantity']<=0).sum()

np.int64(0)

In [31]:
#checking for invalid prices
(df['UnitPrice']<=0).sum()

np.int64(40)

In [32]:
#Removing the invalid rows
df=df[(df['Quantity']>0) & (df['UnitPrice']>0)]


In [34]:
(df['Quantity']<=0).sum() #confirming the removal of invalid quantities
(df['UnitPrice']<=0).sum() #confirming the removal of invalid prices

np.int64(0)

Convert Date and Create Revenue Column

In [38]:
#convert InvoiceDate to datetime
df['InvoiceDate']=pd.to_datetime(df['InvoiceDate'])

In [39]:
df['InvoiceDate'].dtype

dtype('<M8[ns]')

In [40]:
#create revenue column
df['Revenue']=df['Quantity']*df['UnitPrice']

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 397884 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    397884 non-null  object        
 1   StockCode    397884 non-null  object        
 2   Description  397884 non-null  object        
 3   Quantity     397884 non-null  int64         
 4   InvoiceDate  397884 non-null  datetime64[ns]
 5   UnitPrice    397884 non-null  float64       
 6   CustomerID   397884 non-null  int64         
 7   Country      397884 non-null  object        
 8   Revenue      397884 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(4)
memory usage: 30.4+ MB


In [42]:
df[['Quantity','UnitPrice','Revenue']].head()

Quantity  UnitPrice  Revenue
0         6       2.55    15.30
1         6       3.39    20.34
2         8       2.75    22.00
3         6       3.39    20.34
4         6       3.39    20.34